Add runtime for AWS Lambda (#1127)
urbica/martin: docs/src/run-with-lambda.md

[…]
        "repositoryUri": "493749042871.dkr.ecr.us-east-2.amazonaws.com/martin",
# Read the repositoryUri which includes your account number
$ docker tag ghcr.io/maplibre/martin:latest 493749042871.dkr.ecr.us-east-2.amazonaws.com/martin:latest
$ aws ecr get-login-password --region us-east-2 | docker login --username AWS --password-stdin 493749042871.dkr.ecr.us-east-2.amazonaws.com
$ docker push 493749042871.dkr.ecr.us-east-2.amazonaws.com/martin:latest
```

Now you can go to the [Lambda console](https://console.aws.amazon.com/lambda) and create your function.

1. Click “Create function”.
2. Choose “Container image”.
3. Put something in “Function name”. (Note: This is an internal identifier, not exposed in the function URL.)
4. Click “Browse images”, and select your repository and the tag. (If you can’t find it, see if you’re in the same region?)
5. Expand “Container image overrides”, and under CMD put the URL of a .pmtiles file.
6. Set “Architecture” to arm64 to match the platform that we pulled. (Lambda has better ARM CPUs than x86.)
7. Click “Create function”.
8. Find the “Configuration” tab, select “Function URL”, “Create function URL”.

> 9. Set “Auth type” to NONE. (Don’t enable CORS, as Martin already will, and you will have duplicate headers and your CORS will be broken.)

Reviewer: ```suggestion
9. Set “Auth type” to `NONE`
   * Do not enable CORS. Martin already has CORS support, so it will create duplicate headers and break CORS.
```

10. Click on the “Function URL”. If it works, hooray! If it doesn’t, open the “Monitor” tab, “View CloudWatch logs”, find the most recent Log stream.

### Zip deployment

It’s possible to deploy the entire codebase from the AWS console, but we will use Serverless Application Model. Our function will consist of a “Layer”, containing the Martin binary, and our function itself will contain the configuration in .yaml format.

#### The layer

Download the binary and place it in your staging directory. The `bin` directory of your Layer will be added to the PATH.

```bash
mkdir -p martin_layer/src/bin/
cd martin_layer

> curl -OL https://github.com/maplibre/martin/releases/download/VERSION_NUMBER_HERE/martin-aarch64-unknown-linux-musl.tar.gz

Reviewer: Use `latest` instead of requiring version number

```suggestion
curl -OL https://github.com/maplibre/martin/releases/latest/download/martin-aarch64-unknown-linux-musl.tar.gz
```

tar -C src/bin/ -xzf martin-aarch64-unknown-linux-musl.tar.gz martin
```

> Every zip-based Lambda function runs a file called `bootstrap`. `vim src/bootstrap`

Reviewer: better use heredoc for this - easier, and does not require familiarity with `vim`.  Might also use `#!/usr/bin/env sh`, but not certain which shebang is better.

```
cat <<EOF > src/bootstrap
#!/bin/sh
set -eu
exec martin -c ${_HANDLER}.yaml
EOF
```

Author: Looks like AL2023's /bin/sh is "GNU bash, version 5.2.15(1)-release (x86_64-amazon-linux-gnu)", and they also have a merged / and /usr. So it should be fine.

I'm amused to learn that `#!/bin/sh` is actually not a POSIX compliant thing to do, as the path to `sh` cannot be assumed.

Anyway, `/bin/sh` causes Bash to start up in its posix mode, and refers to Dash on Debian and Ubuntu systems, but this small shell script shouldn't encounter any differences :)

```bash
#!/bin/sh
exec martin -c ${_HANDLER}.yaml
```

> Write the SAM template. `vim template.yaml`

Reviewer: use heredoc as above

```yaml
AWSTemplateFormatVersion: 2010-09-09
Transform: 'AWS::Serverless-2016-10-31'
Resources:
  martin:
    Type: 'AWS::Serverless::LayerVersion'
    DeletionPolicy: Delete
    Properties:
      ContentUri: src
      CompatibleRuntimes:
      - provided.al2023
      CompatibleArchitectures:
      - arm64
Outputs:
  LayerArn:
    Value: !Ref MartinLayer
    Export:
      Name: !Sub "${AWS::StackName}-LayerArn"
```

Run `sam deploy --guided`.

> 1. Stack Name: [martin-layer]

Reviewer: broken link?

2. Press enter for everything else
3. The settings are saved to `samconfig.toml`, so you can later do `sam deploy` to update the version, or `sam delete`.

Now if you visit the [Lambda console](https://console.aws.amazon.com/lambda/home) and select “Layers”, you should see your layer.

#### The function

1. Select “Functions”, “Create function”.
2. Put something in “Function name”.
3. Set “Runtime” to “Amazon Linux 2023”.
4. Set “Architecture” to “arm64”.
5. Under “Advanced settings”, choose “Enable function URL” with “Auth type” of “NONE”.
6. Click “Create function”.

Add your layer:

> 1. Click “add a layer” (green banner at the top, or the very bottom).

Reviewer: in cases like these, you may want to add screenshot images, but we can do this in a separate PR.

2. Choose “Custom layers”, and select your layer and its version.
3. Click “Add”.

Add your configuration file in the function source code:

1. Code tab, File, New File: `hello.handler.yaml`.

   ```yaml
   pmtiles:
     sources:
       demotiles: <url to a pmtiles file>
   ```

2. Click Deploy, wait for the success banner, and visit your function URL.

> ### Extra things

Reviewer: "things" here looks weird - either `Extras` or `TODO` ?

```suggestion
### TODO
```

- Lambda has a default timeout of 3 seconds, and 128 MB of memory, maybe this is suboptimal.

> - Martin will cache tiles in memory, but this might be useless for an ephemeral server.

Reviewer: I don't think this is useless - if martin binary will handle more than one request, it will be better to cache at least some things - for example, when using pmtiles, cache the directory structure.
